File: Java/Bellman_Ford.java
import java.io.*;
import java.util.*;

public class Bellman_Ford {
    static class Triad {
        private int source;
        private int target;
        private int weight;
        public Triad(int source, int target, int weight) {
            this.source = source;
            this.target = target;
            this.weight = weight;
        }
        public int getSource() { return this.source; }
        public int getTarget() { return this.target; }
        public int getWeight() { return this.weight; }
    }
    public static boolean BellmanFord(int source, ArrayList<Triad> graph,
            int[] dist, int[] path) {
        int u, v, w;
        dist[source] = 0;
        for (int j=1; j<=dist.length-1; j++) {
            for (int l=0; l<graph.size(); l++) {
                u = graph.get(l).getSource();
                v = graph.get(l).getTarget();
                w = graph.get(l).getWeight();
                if (dist[u] != Integer.MAX_VALUE && (dist[u] + w < dist[v])) {
                    dist[v] = dist[u] + w;
                    path[v] = u;
                }
            }
        }
        for (int i=0; i<graph.size(); i++) {
            u = graph.get(i).getSource();
            v = graph.get(i).getTarget();
            w = graph.get(i).getWeight();
            if (dist[u] != Integer.MAX_VALUE && (dist[u] + w < dist[v])) {
                return false;
            }
        }
        return true;
    }
    public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        ArrayList<Triad> graph = new ArrayList<Triad>();
        int[] dist = new int[n];
        int[] path = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        Arrays.fill(path, -1);
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();
            graph.add(new Triad(u, v, w));
        }
        int s = 0, t = 4;
        boolean res = BellmanFord(s, graph, dist, path);
        if (res == false) {
            System.out.println("Graph contains negative weight cycle");
        }
        else {
            System.out.println(dist[t]);
        }
	}
}

File: Python3/Bellman_Ford.py
import queue
MAX = 100
INF = int(1e9)
class Triad:
    source = 0
    target = 0
    weight = 0
    def __init__(self, source, target, weight):
        self.source = source
        self.target = target
        self.weight = weight

def BellmanFord(source, graph, dist):
    dist[source] = 0
    for i in range(1, n):
        for j in range(m):
            u = graph[j].source
            v = graph[j].target
            w = graph[j].weight
            if dist[u] != INF and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                path[v] = u
    for i in range(m):
        u = graph[i].source
        v = graph[i].target
        w = graph[i].weight
        if dist[u] != INF and dist[u] + w < dist[v]:
            return False
    return True

if __name__ == '__main__':
    n, m = map(int, input().split())
    graph = []
    dist = [INF for i in range(n+5)]
    path = [-1 for i in range(n+5)]
    for i in range(m):
        u, v, w = map(int, input().split())
        graph.append(Triad(u, v, w));
    s, t = 0, 4
    res = BellmanFord(s, graph, dist)
    if res == False:
        print("Graph contains negative weight cycle")
    else:
        print(dist[t])
